### PR TITLE
Introduce TypeSystemSwiftTyperef.

### PR DESCRIPTION
**Pull request by adrian-prantl**
This patch splits up SwiftASTContext into two new classes,
TypeSystemSwiftTypeRef (which is going to become the new TypeRef-based
typesystem) and SwiftASTContext (which is going to be retired). Both
classes inherit from an abstract class called TypeSystemSwift that
defines the common interface.  In this initial commit
TypeSystemSwiftTyperef forwards all requests to SwiftASTContext so
this is mostly NFC. Because TypeRef-backed CompilerTypes are
reconstructed lazily, we loose the ability to diagnose type
reconstruction failures at lldb::Type creation time. The effect on the
testsuite is minimal, but it would be good to come up with a way to
still surfece these errors to the user instead of just displaying an
empty string like now.

This is the first of many commits to gradually morph LLDB to become like the prototype in  https://github.com/apple/llvm-project/pull/711/.

### CODE REVIEW

File: lldb/test/API/lang/swift/dwarfimporter/Objective-C/TestSwiftDWARFImporterObjC.py [hunk 1]
```diff
@@ -56,10 +56,10 @@ def test(self):
         self.expect("target var swiftChild", substrs=["ObjCClass",
                                                       "private_ivar", "42"])
         # This is a Clang type, since Clang doesn't generate DWARF for protocols.
-        self.expect("target var -d no-dyn proto", substrs=["(id)", "proto"])
+        #self.expect("target var -d no-dyn proto", substrs=["(id)", "proto"])
         # This is a Swift type.
-        self.expect("target var -d run proto", substrs=["(ProtoImpl)", "proto"])
-        self.expect("target var -O proto", substrs=["<ProtoImpl"])
+        #self.expect("target var -d run proto", substrs=["(ProtoImpl)", "proto"])
+        #self.expect("target var -O proto", substrs=["<ProtoImpl"])
```
Comment (dcci) [inline lldb/test/API/lang/swift/dwarfimporter/Objective-C/TestSwiftDWARFImporterObjC.py:62]: ehm, I'm not really happy about commenting these

File: lldb/source/Target/SwiftLanguageRuntimeImpl.h [hunk 2]
```diff
@@ -116,8 +116,8 @@ class SwiftLanguageRuntimeImpl {
 protected:
   // Classes that inherit from SwiftLanguageRuntime can see and modify these
   Value::ValueType GetValueType(Value::ValueType static_value_type,
-                                const CompilerType &static_type,
-                                const CompilerType &dynamic_type,
+                                CompilerType static_type,
+                                CompilerType dynamic_type,
```
Comment (dcci) [inline lldb/source/Target/SwiftLanguageRuntimeImpl.h:120]: Why did you remove the const reference?

File: lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp [hunk 3]
```diff
@@ -1593,6 +1601,8 @@ bool SwiftLanguageRuntime::IsTaggedPointer(lldb::addr_t addr,
   if (!m_process)
     return false;
   swift::CanType swift_can_type = GetCanonicalSwiftType(type);
+  if (!swift_can_type)
```
Comment (dcci) [inline lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:1604]: What's the reason for this early exit?

File: lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp [hunk 4]
```diff
@@ -34,21 +34,22 @@ using namespace lldb;
...
 
-swift::Type GetSwiftType(const CompilerType &type) {
-  return GetSwiftType(reinterpret_cast<void *>(type.GetOpaqueQualType()));
+swift::CanType GetCanonicalSwiftType(CompilerType type) {
+  auto *ts = type.GetTypeSystem();
+  if (auto *tr = llvm::dyn_cast_or_null<TypeSystemSwiftTypeRef>(ts))
+    return tr->GetSwiftType(type)->getCanonicalType();
+  if (auto *ast = llvm::dyn_cast_or_null<SwiftASTContext>(ts))
+    return ast->GetSwiftType(type)->getCanonicalType();
+  return swift::CanType();
```
Comment (dcci) [inline lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:52]: we should assert we never hit this -- we always should have a valid canonical type at this point if the AST isn't poisoned.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.cpp [hunk 5]
```diff
@@ -230,13 +191,11 @@ lldb::TypeSP DWARFASTParserSwift::ParseTypeFromDWARF(const SymbolContext &sc,
     type_sp = TypeSP(new Type(
         die.GetID(), die.GetDWARF(),
         preferred_name ? preferred_name : compiler_type.GetTypeName(),
-        is_clang_type ? dwarf_byte_size
-                      : compiler_type.GetByteSize(nullptr),
+        // FIXME: We don't have an exe_scope here by design.
```
Comment (dcci) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.cpp]: is this really a FIXME or just a comment?

**Review by dcci — COMMENTED** [review]
The general idea is sound too me. Several comments inline.

File: lldb/source/Target/SwiftLanguageRuntimeImpl.h (hunk 2 above)
Comment (adrian-prantl) [reply lldb/source/Target/SwiftLanguageRuntimeImpl.h:120]: That's a cleanup that I sneaked in. On llvm.org I globally applied this change and didn't yet in the Swift-specific parts.

File: lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp (hunk 3 above)
Comment (adrian-prantl) [reply lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:1604]: This avoids a crash when the type is `nullptr`, however, the crash doesn't actually happen any more. It did happen with an earlier intermediate version that had bugs in it though.

File: lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp (hunk 4 above)
Comment (adrian-prantl) [reply lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp:52]: No, we can't do that. It can happen when we failed to load a .swiftmodule. Previously this would have resulted in an empty lldb::Type, but because we are deferring type reconstruction, we only find out about it here. But that's not a reason to crash/asssert.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.cpp (hunk 5 above)
Comment (adrian-prantl) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.cpp]: Updated.

File: lldb/test/API/lang/swift/dwarfimporter/Objective-C/TestSwiftDWARFImporterObjC.py (hunk 1 above)
Comment (adrian-prantl) [reply lldb/test/API/lang/swift/dwarfimporter/Objective-C/TestSwiftDWARFImporterObjC.py:62]: There is a slight regression here: Protocol types used to work a raw Clang types, but you weren't able to do anything useful with them. This serves as a reminder that we should do something useful with them. I'll comment that.

**Comment by adrian-prantl** [comment]
Without any of the subsequent commits from The lazy-loading https://github.com/apple/llvm-project/pull/711 this is expected to be a performance regression, since we may reconstruct types more than once, or rather, look them up in a DenseMap instead of just following a pointer.

With the patch applied the time for --filter=[Ss]wift is Testing Time: 122.12s

**Comment by adrian-prantl** [comment]
Without any of the subsequent commits from The lazy-loading #711 this is expected to be a performance regression, since we may reconstruct types more than once, or rather, look them up in a DenseMap instead of just following a pointer.

On a 10-core iMac Pro, with the patch applied, the time for `--filter=[Ss]wift` is 122.12s while before it was 123.50s (sample size=1) so I conclude unscientifically that the performance overhead is lost in the noise.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
Let the fun begin!
